### ISSUE TITLE
How to attach external source-code library to monorepo

### ISSUE THREAD

**Issue by crowmagnumb** [nrwl/nx#3843]
## Description

I have a monorepo that is in turn dependent on a set of libraries, of my own writing, which I have separate because I use them in other projects and some day may open source them if they prove robust and useful enough. So basically, they are not in my monorepo for this reason. BUT I would love to have my monorepo setup such that they are effectively "in it" and I gain the same benefits that I do with my other libraries i.e. changes to them would be instantaneously reflected in my monorepo. To that end I have tried to add a path in my `tsconfig.base.json` file pointing to the source-code outside of my monorepo code (pretty much exactly what you do to reflect the lib/<project> path inside the monorepo). e.g.

```
"paths": {
    "@<myscope>/<my_extern_lib>": ["../<my_external_lib_src_dir>/index.ts"]
}
```

I get a couple of problems right off the bat trying to serve an app dependent on this library.

1) Sass does not seem to recognize this path. I get `SassError: Can't find stylesheet to import` for a line that reads `~@<myscope>/<my_extern_lib>/scss/styles`

2) I get the error `ERROR in Unable to write a reference to <MyComponent> in <my_external_lib_src_dir>/<path_to_component> from <my_external_lib_src_dir>/module.ts`

So angular also is having an issue compiling this.

Any thoughts on how to properly make this work?

## Motivation

It would be super awesome to have this say when you have an issue with a 3rd party library and you create a fork to try and fix the problem. If you could have an easy way to attach the source code from this forked repo into your monorepo it would make fixing and submitting pull requests a breeze.

## Suggested Implementation

So ultimately an awesome thing would be ...

```
nx generate external_lib <name> <path>
```

**Comment by FrozenPandaz** [comment]
I think what you are looking for might be a combination of `git submodules` and `npm link`?

* https://www.git-scm.com/book/en/v2/Git-Tools-Submodules
* https://docs.npmjs.com/cli/link

I have not done it myself so I do not know how well it would work but in _theory_ you could have a submodule in your monorepo... which contains the external project which you could build and use `npm link` to link your monorepo to the outputs of the git submodule?

You'd lose atomic commits since you'd have to commit to the submodule and then bring that submodule's commit up to date in your monorepo.. but it could suit your needs?

For a 3rd party library.. I would avoid using a submodule if it's temporary. 

It's not clear how we would bring this kind of feature into Nx or if it is in the scope of Nx though.

**Comment by crowmagnumb** [comment]
What I decided to do in the end is just give up on keeping those other libraries in a separate repo because I've been fighting with this too long. I will just pull them out again later if I need to and have the time. For right now I just need to move forward with my normal work. The problem for me here has been the move to Angular 10 and Ivy combined with NativeScript 7. Both of these technologies I love but this has been a really, really tough upgrade between all these diifferent things trying to work together along with ng-packagr. Everything I try to do with packaging these libs (Ivy or not) leads to some kind of problem. So i'm ditching that all for now and going full on monorepo.

Just for anyone reading this in the future, I did discover, only after merging the last libs into my monorepo, that I probably could have solved the scss import issue with ["stylePreprocessorOptions"](https://github.com/angular/angular-cli/issues/9018) added to my `angular.json` file.

Great product! I'm so happy to be an `nrwl` train. :)

**Comment by agentlewis** [comment]
Hey @crowmagnumb. Just wanted to share that we have a similar requirement and I have been looking at various solutions.

We are a consulting firm and for most clients end up needing to publish several "apps". Each client also has at least 1 library we create for them. 

We then in turn have several libraries we have developed in house that we use across client projects (api adapters and runtime validation libraries). Ideally we want to be able to add-hoc develop these while we are also working with our client solutions. 

The ideal would be if NX allowed for a lib to have its own github repo, package.json and build/publish command. Essentially allowing us to pull it in and work with in within the nx solution - but publish out changes as needed. I could see how this could be a git-submodule solution. But I have been down the submodule path before and learned that its very hard to have a team of devs working with it.

As such we are now looking to use [meta](https://github.com/mateodelnorte/meta) with yarn workspaces as a foundation for each client. Into this we will add our cross-client packages (and repos) and a new nx solution. 

Something like:

```
meta-clientName (github repo) - created from github template. inc. yarn/npm workspaces for efficient installs of libs. 
 - customLibrary1 (github repo - published to github private package)
 - customLibrary2 (github repo - published to github private package)
 - nx-clientName (github repo) - created from cli (plus changes to support [firebase functions](https://github.com/nrwl/nx/issues/836#issuecomment-610396769))
  - .npmrc file to access published private packages
  - apps
   - clientApp1
   - clientApp2
  - libs
   - clientLib1 (published to github private package)
   - clientLib2 (published to github private package)
 - *clientNameIsolated (github repo) - created from a github template
  - .npmrc file to access private packages inc. clientLib1 & clientLib2
```

* sometimes we will want to bring a contractor on to work on a specific thing and not want them to have access to all the code - this could be for security or we don't want to ask them to learn NX, Typescript or something else. In this case we will make another client repo in meta and bring the dev into that specific work space. Once they were done we would likely migrate the code into nx as an app.

With this we are hoping the DX of developing client solutions with cross client packages will improve.

Anyway I haven't tested this yet - I need to first finish refactoring from lerna to NX. But this plan is feeling pretty solid for now.

cc @iainkirkpatrick

**Comment by dobromyslov** [comment]
@agentlewis thank you for sharing this approach. I also was looking for some solution to mix a private NX workspace monorepo with my external open source libraries located in separate github repositories.

And found old and bold solution from @vsavkin https://blog.nrwl.io/dev-workflow-using-git-submodules-and-yarn-workspaces-14fd06c07964 but it was posted 3 years ago. Also folks say it does not work properly now.

I would be glad to hear from you any success or failure about meta integration into the dev pipeline. Do you have any news?

P.S. Have a look at this improved NX Firebase support workaround https://github.com/nrwl/nx/issues/836#issuecomment-714384504

**Comment by dobromyslov** [comment]
@agentlewis tried to use npm@7 workspaces with no luck. It's yet very raw and unstable.
Implemented your schema with `meta` and `meta-npm` plugin.

```
meta npm install
meta npm build
meta npm link
```

And having troubles with autoimport feature in WebStorm. Looks like it does not work well with this project structure and tries to import from libs source code by their relative paths but not by their package names.

Will give a try to yarn workspaces when get some spare time.

**Comment by crowmagnumb** [comment]
@agentlewis Hmmm. `meta` does indeed seem promising. I'll have to take a look at that when I have some time, but sure would like to see that auto-integrated here.  :)

**Comment by jebesampaio** [comment]
@agentlewis did you manage to get that setup working (with meta)? I'm still struggling with this issue...

**Comment by agentlewis** [comment]
@jebesampaio (and others that pinged me - sorry I never replied)

So in the end we did not use meta - just yarn workspaces and github rep templates.

We also used a seperate yarn workspace for publishing the private npm packages that were used across several projects. These are published to Github as private node packages.

This worked ok, but did not become the "system" we could standardise across the company.

My business partner and far more technically read friend @iainkirkpatrick has just done a new push for our business on creating a mono-repo template using [TurboRepo](https://turborepo.org/) which having just seen his work, looks very promising.

I could see this working in tandem with repo templates and possibly still a seperate project with cross project libraries existing being published from a different repo.

So basically we are not using nx any longer :( 

I think there are features in nx we would like to have, but stanardising via repo templates and shared published packages has ended up being a simpler solution for us.
